backend/src/crd/controller/viewer/reconciler/reconciler.go

```diff
@@ -0,0 +1,224 @@
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const viewerTargetPort = 6006
+
+// Reconciler implements reconcile.Reconciler for the Viewer CRD.
+type Reconciler struct {
+	client.Client
+	scheme *runtime.Scheme
+}
+
+// New returns a new Reconciler.
+func New(cli client.Client, scheme *runtime.Scheme) *Reconciler {
+	return &Reconciler{Client: cli, scheme: scheme}
+}
+
+// Reconcile runs the main logic for reconciling the state of a viewer with a
+// corresponding deployment and service allowing users to access the view under
+// a specific path.
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	glog.Infof("Reconcile request: %+v", req)
+
+	view := &viewerV1alpha1.Viewer{}
+	if err := r.Get(context.Background(), req.NamespacedName, view); err != nil {
+		if errors.IsNotFound(err) {
+			// No viewer instance, so this may be the result of a delete.
+			// Nothing to do.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+	glog.Infof("Got instance: %+v", view)
+
+	// Ignore other viewer types for now.
+	if view.Spec.Type != viewerV1alpha1.ViewerTypeTensorboard {
+		glog.Infof("Unsupported spec type: %q", view.Spec.Type)
+		// Return nil to indicate nothing more to do here.
+		return reconcile.Result{}, nil
+	}
+
+	// Set up potential deployment.
+	dpl, err := deploymentFrom(view)
+	if err != nil {
+		utilruntime.HandleError(err)
+		// User error, don't requeue key.
+		return reconcile.Result{}, nil
+	}
+
+	// Set the deployment to be owned by the view instance. This ensures that
+	// deleting the viewer instance will delete the deployment as well.
+	if err := controllerutil.SetControllerReference(view, dpl, r.scheme); err != nil {
+		// Error means that the deployment is already owned by some other instance.
+		utilruntime.HandleError(err)
+		return reconcile.Result{}, err
+	}
+
+	foundDpl := &appsv1.Deployment{}
+	nsn := types.NamespacedName{Name: dpl.Name, Namespace: dpl.Namespace}
+	if err := r.Client.Get(context.Background(), nsn, foundDpl); err != nil {
+		if errors.IsNotFound(err) {
+			// Create a new instance.
+			if err := r.Client.Create(context.Background(), dpl); err != nil {
+				utilruntime.HandleError(fmt.Errorf("error creating deployment: %v", err))
+				return reconcile.Result{}, err
+			}
+		} else {
+			// Some other error.
+			utilruntime.HandleError(err)
+			return reconcile.Result{}, err
+		}
+	}
+	glog.Infof("Created new deployment with spec: %+v", dpl)
+
```

Comment:
Is there a way to write simple fakes like in the case of the persistence agent: 

https://github.com/kubeflow/pipelines/tree/master/backend/src/agent/persistence/client

So that we can test end to end?

Reply:
Yes, there is. This would be what we'd use:
https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/reconcile/reconciletest

I will add an e2e test in a follow up PR. For now, I added unit tests testing the behaviour of the reconciler using the fake kubernetes client here:
https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/client/fake